File: dailyQuestion/2020-07/07-18/golang/solution_roll_array.go
package main

import "fmt"

func isInterleave(s1 string, s2 string, s3 string) bool {
	m, n, t := len(s1), len(s2), len(s3)
	if m+n != t {
		return false
	}
	dp := make([]bool, n+1)
	dp[0] = true
	for i := 0; i <= m; i++ {
		for j := 0; j <= n; j++ {
			if i > 0 {
				dp[j] = (dp[j] && s1[i-1] == s3[i+j-1])
			}
			if j > 0 {
				dp[j] = dp[j] || (dp[j-1] && s2[j-1] == s3[i+j-1])
			}
		}
	}
	return dp[n]
}
func main() {
	s1 := "aabcc"
	s2 := "dbbca"
	s3 := "aadbbbaccc"
	result := isInterleave(s1, s2, s3)
	fmt.Println(result)
}

File: dailyQuestion/2020/2020-07/07-18/python/solution_roll_array.py
def isInterleave(s1, s2, s3):
    """
    :type s1: str
    :type s2: str
    :type s3: str
    :rtype: bool
    """
    m, n, t = len(s1), len(s2), len(s3)
    if m + n != t:
        return False
    dp = [False] * (n + 1)
    dp[0] = True
    for i in range(m + 1):
        for j in range(n + 1):
            if i > 0:
                dp[j] = dp[j] and s1[i - 1] == s3[i + j - 1]
            if j > 0:
                dp[j] = dp[j] or (dp[j - 1] and s2[j - 1] == s3[i + j - 1])
    return dp[n]

if __name__ == "__main__" :
    s1 = "aabcc"
    s2 = "dbbca"
    s3 = "aadbbbaccc"
    result = isInterleave(s1,s2,s3)
    print(result)

